shared/lib/accounts/accounts.ts

```diff
@@ -35,46 +51,199 @@ export function getUniqueAccountName(
+    case SOLANA_WALLET_SNAP_ID: {
+      // Solana accounts should have in their scope the 3 networks
+      // mainnet, testnet, and devnet. Therefore, we can use this name
+      // for all 3 networks.
+      return `Solana Account ${accountNumber}`;
+    }
+    default:
+      return defaultSnapAccountName;
+  }
+}
+
+///: BEGIN:ONLY_INCLUDE_IF(multichain)
+export type WalletSnapOptions = {
+  scope?: CaipChainId;
+  derivationPath?: DiscoveredAccount['derivationPath'];
+  entropySource?: string;
+  accountNameSuggestion?: string;
+};
+
+export type WalletSnapClient = {
+  getSnapId(): SnapId;
+
+  createAccount(
+    options: WalletSnapOptions,
+    internalOptions?: SnapKeyringInternalOptions,
+  ): Promise<KeyringAccount>;
+
+  getNextAvailableAccountName(
+    options?: SnapAccountNameOptions,
+  ): Promise<string>;
+};
+
+export type MultichainWalletSnapClientMessenger = Messenger<
+  | SnapControllerHandleRequest
+  | AccountsControllerGetNextAvailableAccountNameAction,
+  never
+>;
+
+export class MultichainWalletSnapClient implements WalletSnapClient {
+  readonly #snapId: SUPPORTED_WALLET_SNAP_ID;
+
+  readonly #snapKeyring: SnapKeyring;
+
+  readonly #scopes: CaipChainId[];
 
-    await Promise.allSettled(
-      discovered.map(async (discoveredAccount) => {
-        const options = {
-          derivationPath: discoveredAccount.derivationPath,
-          entropySource,
-        };
-
-        // TODO: Use `withKeyring` instead of using the keyring directly.
-        await snapKeyring.createAccount(snapId, options, {
-          displayConfirmation: false,
-          displayAccountNameSuggestion: false,
-          setSelectedAccount: false,
-        });
+  readonly #client: KeyringInternalSnapClient;
+
+  readonly #messenger: MultichainWalletSnapClientMessenger;
+
+  constructor(
+    snapId: SUPPORTED_WALLET_SNAP_ID,
+    scopes: CaipChainId[],
+    snapKeyring: SnapKeyring,
+    messenger: MultichainWalletSnapClientMessenger,
+  ) {
+    this.#snapId = snapId;
+    this.#snapKeyring = snapKeyring;
+
+    this.#scopes = scopes;
+
+    this.#messenger = messenger;
+
+    this.#client = new KeyringInternalSnapClient({
+      snapId,
+      // @ts-expect-error TODO: Resolve mismatch between base-controller versions.
+      messenger: messenger.getRestricted({
+        name: 'KeyringInternalSnapClient',
+        allowedActions: ['SnapController:handleRequest'],
+        allowedEvents: [],
       }),
+    });
+  }
+
+  getSnapId(): SnapId {
+    return this.#snapId;
+  }
+
+  async createAccount(
+    { accountNameSuggestion, ...options }: WalletSnapOptions,
+    internalOptions?: SnapKeyringInternalOptions,
+  ): Promise<KeyringAccount> {
+    // Automatically name the account if not provided.
+    const autoAccountNameSuggestion = accountNameSuggestion
+      ? { accountNameSuggestion }
+      : { accountNameSuggestion: await this.getNextAvailableAccountName() };
+
+    // TODO: Use `withKeyring` instead of using the keyring directly.
```

Comment:
Hopefully there is an issue capturing these TODO items.

Reply:
Yup, I did add some tickets to keep track of this 👍